code/modules/reagents/chemistry/machinery/reagentgrinder.dm

```diff
@@ -520,11 +521,20 @@
 		if(beaker.reagents.holder_full())
 			return
 
-	// Everything else - Transfers reagents from it into beaker
-	for(var/obj/item/reagent_containers/O in holdingitems)
-		O.reagents.trans_to(beaker, O.reagents.total_volume)
+	// Cigarettes
+	for(var/obj/item/clothing/mask/cigarette/O in holdingitems)
+		if(grind_items(O))
+			return
 
-		if(!O.reagents.total_volume)
-			remove_object(O)
-		if(beaker.reagents.holder_full())
+	// Everything else
+	for(var/obj/item/reagent_containers/O in holdingitems)
+		if(grind_items(O))
 			return
+
+/// Will transfer reagents of the item to the beaker, and returns `TRUE` if the beaker is full
+/obj/machinery/reagentgrinder/proc/grind_items(obj/item/O)
+	O.reagents.trans_to(beaker, O.reagents.total_volume)
+	if(!O.reagents.total_volume)
+		remove_object(O)
+	if(beaker.reagents.holder_full())
+		return TRUE
```

Comment:
This can all be simplified into a single look that goes over all `/obj/item` in the items

Reply:
Done.